source/Mission.cpp

```diff
@@ -1068,14 +1080,27 @@ Mission Mission::Instantiate(const PlayerInfo &player) const
 
 
 
+// Perform an "on enter" MissionAction associated with the current system.
 void Mission::Enter(const System *system, PlayerInfo &player, UI *ui)
 {
-	auto eit = onEnter.find(system);
-	if(eit != onEnter.end() && !didEnter.count(eit->first))
+	const auto &eit = onEnter.find(system);
+	if(eit != onEnter.end() && !didEnter.count(eit->first)
+			&& eit->second.CanBeDone(player))
 	{
 		eit->second.Do(player, ui);
 		didEnter.insert(eit->first);
 	}
+	// If this system has not yet matched to an "on enter" action for this mission,
+	// consider matching a generic "on enter" to it. A generic "on enter" action
+	// may have a LocationFilter that governs which systems it can be performed in.
+	if(!genericOnEnter.empty() && !didEnter.count(system))
+		for(auto it = genericOnEnter.begin(); it != genericOnEnter.end(); ++it)
+			if((*it).CanBeDone(player))
+			{
+				(*it).Do(player, ui);
+				genericOnEnter.erase(it);
+				break;
+			}
```

Comment:
The reason for "didEnter" is that if you erase the MissionAction immediately, and the MissionAction contains a conversation, that conversation is deleted while the panel that displays it is still active, resulting in a crash or other weird behavior.

A better option would be to modify didEnter to store a pointer to a MissionAction rather than to a system. That way, it could apply to both the sets of one enter actions, e.g.:

```
if(eit != onEnter.end() && !didEnter.count(&eit->second)...
```

Also, if multiple "on enters" apply to the same system, I feel like it's better to apply all of them (showing multiple dialogs, if necessary) rather than allowing all but one of them to silently be ignored.

Reply:
As written the generic `on enter`s wouldn't be ignored, just not activated (since some other O.E. activated instead). The next time they match a system they would be activated (since I was using the direct erasure and not checking for them in `didEnter`).
Without getting crazy specific in a LocationFilter, it's difficult to set a "sequential" scene, where a conversation triggers in one system and then a related conversation triggers in the next. With a single generic O.E. triggering per system entry (and only if a specific O.E. did not), this is a lot easier to do, and the order of display is controlled by the order of specification in the datafile.

I'll make the didEnter<MA*> change and we can invite further discussion on displaying all matched generics vs only one per jump.